### DISCUSSION TITLE
Deprecated `socket_options`

### DISCUSSION THREAD

**Discussion by karpetrosyan** [encode/httpcore#800] [General]
#668 added support for socket options.
Refs: https://github.com/encode/httpx/discussions/2811

Because we have a `network_backend` as a public API, we can now drop `socket_options` support and just use a custom `network_backend` instead.

I suggest that we will deprecate `socket_options` and remove it in httpcore 0.19.0

**Comment by lovelydinosaur** [comment]
Ah yes I think you've got a good bit of neatening up here.

To clarify first, could you provide before/after examples?...

**Reply by karpetrosyan** [reply]
Of course

Before:

```python
import socket
import httpcore


pool = httpcore.ConnectionPool(
    socket_options=[(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)]
)
```

After:

```python
import socket
import httpcore


class ReuseaddrNetworkBackend(httpcore.SyncBackend):
    def connect_tcp(
        self,
        host: str,
        port: int,
        timeout: float | None = None,
        local_address: str | None = None,
    ) -> httpcore.NetworkStream:
        stream = super().connect_tcp(host, port, timeout, local_address)
        stream.get_extra_info("socket").setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        return stream


pool = httpcore.ConnectionPool(network_backend=ReuseaddrNetworkBackend())
```

**Reply by lovelydinosaur** [reply]
This isn't quite what I'd expected you to propose here.
Also notable is that we pushed `socket_options` support through to [the `httpx` API](https://github.com/encode/httpx/blob/master/CHANGELOG.md#added).

I think if we wanted any changes here we'd need a clear design discussion on it first.

**Reply by karpetrosyan** [reply]
> This isn't quite what I'd expected you to propose here.

I don't think I understand what you mean here.

**Reply by lovelydinosaur** [reply]
Okay I'll rephrase...

Why do we want to deprecate `socket_options`?

**Reply by karpetrosyan** [reply]
We have a powerful `network_backend` API, which is exactly what we want in such cases. Why should we support additional APIs for doing the same thing?

**Reply by lovelydinosaur** [reply]
*Note: I do think we unneccessarily rushed `socket_options` support into the API, yes. However it is there now, so if we want to change that we need to work through things carefully so as to not negatively impact our users.*

Given that https://github.com/encode/httpx/pull/2716 adds support for `socket_options` to `httpx`, are you also suggesting that we drop support for `httpx.HTTPTransport(socket_options=...)`?

**Reply by karpetrosyan** [reply]
For now, I just suggest deprecating it in httpcore. I think we are isolated enough to work on such things without thinking about HTTPX.

**Reply by karpetrosyan** [reply]
I think we just should avoid removing socket_options in 0.18.* versions (because now httpx depends on httpcore>=18 and < 19 versions), so we can deprecate it in 0.18.1 and move to 0.19 without any problems.
